EventListener/TagSubscriber.php

```diff
@@ -86,8 +113,13 @@ public function onKernelResponse(FilterResponseEvent $event)
         }
 
         if ($request->isMethodSafe()) {
-            // For safe requests (GET and HEAD), set cache tags on response
-            $this->cacheManager->tagResponse($response, $tags);
+            if ($event->isMasterRequest()) {
+                // For safe requests (GET and HEAD), set cache tags on response
+                $this->cacheManager->tagResponse($response, $tags);
+            } else {
+                // Retain tags to attach to response to master request
+                $this->tags = $tags;
+            }
```

Comment:
i ended up doing this logic. otherwise tags from subrequests get lost. is this sane?

for ESI requests, everything is a separate master request and its own cache lifecycle, so having different tags makes sense.